index.js

```diff
@@ -482,6 +482,120 @@ app.post('/api/heartbeat', Validator.body(['coreid', 'event', 'data']).exists(),
+        let client
+
+        try {
+          client = await db.beginTransaction()
+          if (client === null) {
+            helpers.logError(`alertSessionChangedCallback: Error starting transaction`)
+            return
+          }
+
+          const session = await db.getUnrespondedSessionWithLocationId(location.locationid, client)
+          if (session) {
+            session.respondedAt = await db.getCurrentTime(client)
+            await db.saveSession(session, client)
+          } else {
+            helpers.logError(`error stopping session and chatbot due to siren button press`)
+          }
+
+          await db.commitTransaction(client)
+        } catch (e) {
+          try {
+            await db.rollbackTransaction(client)
+            helpers.logError(`alertSessionChangedCallback: Rolled back transaction because of error: ${e}`)
+          } catch (error) {
+            // Do nothing
+            helpers.logError(`alertSessionChangedCallback: Error rolling back transaction: ${e}`)
+          }
+        }
+        response.status(200).json('OK')
+      }
+    } else {
+      const errorMessage = `Bad request to ${request.path}: ${validationErrors.array()}`
+      helpers.logError(errorMessage)
+      // Must send 200 so as not to be throttled by Particle (ref: https://docs.particle.io/reference/device-cloud/webhooks/#limits)
+      response.status(200).json(errorMessage)
+    }
+  } catch (err) {
+    const errorMessage = `Error calling ${request.path}: ${err.toString()}`
+    helpers.logError(errorMessage)
+    // Must send 200 so as not to be throttled by Particle (ref: https://docs.particle.io/reference/device-cloud/webhooks/#limits)
+    response.status(200).json(errorMessage)
+  }
+})
+
+app.post('/api/siren-escalate', Validator.body(['coreid', 'event', 'data']).exists(), async (request, response) => {
+  try {
+    const validationErrors = Validator.validationResult(request).formatWith(helpers.formatExpressValidationErrors)
+
+    if (validationErrors.isEmpty()) {
+      const coreId = request.body.coreid
+      const location = await db.getLocationFromParticleCoreID(coreId)
+
+      if (!location) {
+        const errorMessage = `Bad request to ${request.path}: no location matches the coreID ${coreId}`
+        helpers.logError(errorMessage)
+        // Must send 200 so as not to be throttled by Particle (ref: https://docs.particle.io/reference/device-cloud/webhooks/#limits)
+        response.status(200).json(errorMessage)
+      } else {
+        let client
+
+        try {
+          client = await db.beginTransaction()
+          if (client === null) {
+            helpers.logError(`alertSessionChangedCallback: Error starting transaction`)
+            return
+          }
+
+          const session = await db.getUnrespondedSessionWithLocationId(location.locationid, client)
+          if (session) {
+            session.chatbotState = CHATBOT_STATE.STARTED
+            await db.saveSession(session, client)
```

Comment:
To start a regular session, it would need to call `braveAlerter.startAlertSession()` like in `handleAlert`, but I'm reconsidering this as a good fallback strategy because calling `startAlertSession` will create a new session in the `sessions` table, whereas you already have a session. Maybe you should instead call `braveAlerter.sendSingleAlert` to send a single text message to the fallback number and you can update the session state to `WAITING_FOR_REPLY`.

On the firmware side of things, you're planning to shut of the noise and lights when this function returns no matter what, right?

Reply:
I'll wait for @mariocimet 's feedback on this as I am still relatively unfamiliar with the code base. On the firmware side of things, yes, the siren shuts off after the timer runs out (modifiable to a variety of times).